src/Conversion/ONNXToKrnl/Math/MatMulInteger.cpp

```diff
@@ -0,0 +1,112 @@
+    : public OpConversionPattern<ONNXMatMulIntegerOp> {
+public:
+  ONNXMatMulIntegerOpLowering(TypeConverter &typeConverter, MLIRContext *ctx)
+      : OpConversionPattern(typeConverter, ctx) {}
+
+  LogicalResult matchAndRewrite(ONNXMatMulIntegerOp mmiOp,
+      ONNXMatMulIntegerOpAdaptor adaptor,
+      ConversionPatternRewriter &rewriter) const final {
+    using LocalDialectBuilder = MultiDialectBuilder<KrnlBuilder,
+        IndexExprBuilderForKrnl, MathBuilder, MemRefBuilder, OnnxBuilder>;
+    Operation *op = mmiOp.getOperation();
+    Location loc = ONNXLoc<ONNXMatMulIntegerOp>(op);
+    LocalDialectBuilder create(rewriter, loc);
+
+    ValueRange operands = adaptor.getOperands();
+    Value A = adaptor.getA();
+    Value B = adaptor.getB();
+    Value aZeroPoint = mmiOp.getAZeroPoint(); // Optional input.
+    Value bZeroPoint = mmiOp.getBZeroPoint(); // Optional input.
+
+    if (!isNoneValue(aZeroPoint)) {
+      ShapedType stype = aZeroPoint.getType().dyn_cast<ShapedType>();
+      if ((stype.getRank() > 1) ||
+          (stype.getRank() == 1 && (stype.getShape()[0] != 1))) {
+        return rewriter.notifyMatchFailure(op, [&](mlir::Diagnostic &diag) {
+          diag << "Does not support non-scalar for a_zero_point";
+        });
+      }
+    }
+
+    // Common types.
+    auto resMemRefType = dyn_cast<MemRefType>(
+        typeConverter->convertType(mmiOp.getResult().getType()));
+    Type resElementType = resMemRefType.getElementType();
+
+    // Get shape.
+    ONNXMatMulIntegerOpShapeHelper shapeHelper(op, operands, &create.krnlIE);
+    shapeHelper.computeShapeAndAssertOnFailure();
+
+    Type i32Ty = rewriter.getIntegerType(32);
+    Type ui32Ty = rewriter.getIntegerType(32, /*isSigned=*/false);
+    assert(resElementType == i32Ty && "Output type is not i32");
+
+    Value AUInt32, BUInt32;
+    // Use i32 for onnx.Sub because onnx.Sub does not support ui32.
+    // It's because `arith` dialect does not have `subui` though it has `addui`
+    // and `mului`.
+    if (!isNoneValue(aZeroPoint)) {
+      Value AInt32 = create.onnx.cast(A, i32Ty);
+      Value aZeroPointInt32 = create.onnx.cast(aZeroPoint, i32Ty);
+      // Note: `sub` is not true if aZeroPoint shape is [M], M != 1,
+      // because [MxK] - [M] cannot be expressed by `sub`.
+      // We require that aZeroPoint is scalar here (checked at the beginning of
+      // the lowering).
+      AInt32 = create.onnx.sub(AInt32, aZeroPointInt32);
+      AUInt32 = create.onnx.cast(AInt32, ui32Ty);
+    } else
+      AUInt32 = create.onnx.cast(A, ui32Ty);
+
+    if (!isNoneValue(bZeroPoint)) {
+      Value BInt32 = create.onnx.cast(B, i32Ty);
+      // K is the broadcating dim: [KxN] - [N] = [KxN] - [1xN]
+      Value bZeroPointInt32 = create.onnx.cast(bZeroPoint, i32Ty);
+      BInt32 = create.onnx.sub(BInt32, bZeroPointInt32);
+      BUInt32 = create.onnx.cast(BInt32, ui32Ty);
+    } else
+      BUInt32 = create.onnx.cast(B, ui32Ty);
+
+    // Emit MatMul for ui32.
```

Comment:
No particular reason as MatMulInteger's input is i8 or ui8, so I used unsigned integer here. Perhaps, we should check input's type to decide which one to use. For example, 
- Both A and B is i8 -> use i32
- Both A and B is ui8 -> use ui32
- A is i8, B is ui8 (and vice versa) -> not sure which one is better: i32 or ui32. Looks like i32 is convenient.

What do you think?

Reply:
I recommend to always use i32 because the ranges of i8 and u8 (after taking zero inputs into account) both fit in i32, whereas u32 cannot represent negative values, and not that even u8 input can express negative values if a non-zero zero input is given

the spec also says i32 and that's what I used for constant propagation

Reply:
It makes sense. I moved to using i32 for computation. Thanks for the discussion!